src/shaders/_prelude_fog.fragment.glsl

```diff
@@ -41,17 +56,32 @@ float fog_opacity(vec3 pos) {
     return falloff * u_fog_opacity;
 }
 
+vec3 linear_to_srgb(vec3 color) {
+    return pow(color, vec3(1.0 / 2.2));
+}
+
+vec3 srgb_to_linear(vec3 color) {
+    return pow(color, vec3(2.2));
+}
+
 // Assumes z up
 vec3 fog_apply_sky_gradient(vec3 camera_ray, vec3 sky_color) {
     return mix(sky_color, u_fog_color, fog_sky_blending(normalize(camera_ray)));
 }
 
 vec3 fog_apply(vec3 color, vec3 pos) {
-    // We mix in sRGB color space. sRGB roughly corrects for perceived brightness
-    // so that dark fog and light fog obscure similarly for otherwise identical
-    // parameters. If we blend in linear RGB, then the parameters to control dark
-    // and light fog are fundamentally different.
-    return mix(color, u_fog_color, fog_opacity(pos));
+    float haze_opac = fog_opacity(pos);
+    // When haze is present, raise the fog to a power to decrease it
+    float fog_opac = pow(haze_opac, u_fog_exponent);
+    vec3 haze = u_haziness * haze_opac * u_haze_color_linear;
+
+    // When there's any haze, we prefer the tone map, but when haze goes away,
+    // we transition to simply the original color
+    color = srgb_to_linear(color);
+    color = mix(color, tonemap(color + haze), sqrt(haze_opac));
+
+    // Mix the fog color to the result so that it still blends with the sky
+    return mix(linear_to_srgb(color), u_fog_color, fog_opac);
```

Comment:
For this code path, we may investigate adding a variant for haze disabled (by adding an extra define FOG_HAZE) that would make this code path the simplified expression `return mix(color, u_fog_color, fog_opacity(pos));` when haze == 0.

This may seem extraneous, but since we're adding these new shader statements on every possible rendered fragment, there may be a non-negligible cost to these new function calls (sqrt and pow may be expensive function evaluations).

Reply:
You read my mind. I definitely support this since it's so easy and probably does result in some small but measurable performance boost.